src/internal.c

```diff
@@ -218,6 +218,182 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
     int tsip_generatePremasterSecret();
     int tsip_generateEncryptPreMasterSecret(WOLFSSL *ssl, byte *out, word32 *outSz);
 #endif
+#if defined(OPENSSL_EXTRA) && defined(HAVE_SECRET_CALLBACK)
+
+    static void ToHexString(const byte* bin, byte* hex, int BinSz);
+    static int  SessionSecret_callback(WOLFSSL* ssl, void* secret,
+                                                  int* secretSz, void* ctx);
+    static int  SessionSecret_callback_Tls13(WOLFSSL* ssl, int id,
+                       const unsigned char* secret, int secretSz, void* ctx);
+
+    /* Convert input byte array to Hex string then add a null-terminator
+     *
+     * The buffer for hex must be equal or grater than (bin size * 2 + 1).
+     *
+     */
+    static void ToHexString(const byte* bin, byte* hex, int BinSz)
+    {
+        if (bin == NULL || hex == NULL || BinSz < 2)
+            return;
+
+        byte HexTbl[16] = { '0','1','2','3','4','5','6','7',
+                            '8','9','a','b','c','d','e','f' };
+
+        byte UpNibble, LwNibble;
+
+        for (int i = BinSz; i > 0; i--) {
+            UpNibble = (*bin >> 4) & 0x0f;
+            LwNibble =  *bin & 0x0f;
+
+            *hex++ = HexTbl[UpNibble];
+            *hex++ = HexTbl[LwNibble];
+            bin++;
+        }
+        *hex = '\0';
+    }
+    /*
+     * This API is an implementation of SessionSecretCB to be passed
+     * to wolfSSL_set_session_secret_cb()
+     *
+     */
+    static int SessionSecret_callback(WOLFSSL* ssl, void* secret,
+                    int* secretSz, void* ctx)
+    {
+        (void)ctx;
+
+        if (ssl == NULL || secret == NULL || *secretSz == 0)
+            return BAD_FUNC_ARG;
+
+        /* get the user-callback func from CTX*/
+        wolfSSL_CTX_keylog_cb_func callback = ssl->ctx->Keylog_cb;
+        if (callback == NULL)
+            return 0;
+
+        /* need to make sure the given master-secret has a meaningful value */
+        int mssz   = *secretSz;
+        int hasVal = 0;
+        for (int i = 0; i < mssz; i++) {
+            if (*((byte*)secret) != 0) {
+                hasVal = 1;
+                break;
+            }
+        }
+        if (hasVal == 0)
+            return 0; /* return sd success */
```

Comment:
What does "sd" mean here?  Please make comment more clear, thanks.

Reply:
It was a typo. Changed comment.